src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -894,6 +894,38 @@ public T WithB2CAuthority(string authorityUri)
             return (T)this;
         }
 
+        /// <summary>
+        /// Adds a known authority corresponding to a generic OpenIdConnect Provider (OP).
+        /// See https://openid.net/specs/openid-connect-core-1_0.html#Terminology
+        /// </summary>
+        /// <param name="authorityUri">OpenIdConnect authority</param>
+        /// <param name="discoveryDocumentAddress">Address of the discovery endpoint. See https://openid.net/specs/openid-connect-discovery-1_0.html</param>
+        /// <returns>The builder to chain the .With methods</returns>
+        public T WithGenericAuthority(string authorityUri, string discoveryDocumentAddress = null)
```

Comment:
@jmprieur for public API guidance. 

Context: https://developers.google.com/identity/openid-connect/openid-connect - they suggest that developers hardcode the well-known endpoint in their apps. 

So how should we model this API? Here's what I'm thinking 

## Basic Scenario 

- `WithGenericAuthority(string wellKnownEndpoint)`  

and here MSAL does OIDC discovery and pulls the token endpoint and the authorization endpoint. Internally, for caching, we can say that the `environment` is `wellKnownEndpoint.TrimEnd("/.well-known/openid-configuration")`.

## Advanced scenario

(to be implemented later)

- `WithGenericAuthority(GenericAuthorityInfo.FromOidcDocument(oidcDiscoveryDoc))`  
- `WithGenericAuthority(GenericAuthorityInfo.FromEndpoints(tokenEndpoint, authorizationEndpoint))`  


Thoughts?

Reply:
Please see also https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationRequest that describes that metadata document is supposed to be issuer + "/.well-known/openid-configuration"